staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -0,0 +1,745 @@
+		deviceMatchesRequest: make(map[matchKey]bool),
+		constraints:          make([][]constraint, len(a.claimsToAllocate)),
+		requestData:          make(map[requestIndices]requestData),
+		allocated:            make(map[DeviceID]bool),
+		result:               make([]*resourceapi.AllocationResult, len(a.claimsToAllocate)),
+	}
+
+	// First determine all eligible pools.
+	pools, err := GatherPools(ctx, alloc.sliceLister, node)
+	if err != nil {
+		return nil, fmt.Errorf("gather pool information: %w", err)
+	}
+	alloc.pools = pools
+	if loggerV := alloc.logger.V(6); loggerV.Enabled() {
+		loggerV.Info("Gathered pool information", "numPools", len(pools), "pools", pools)
+	} else {
+		alloc.logger.V(4).Info("Gathered pool information", "numPools", len(pools))
+	}
+
+	// We allocate one claim after the other and for each claim, all of
+	// its requests. For each individual device we pick one possible
+	// candidate after the other, checking constraints as we go.
+	// Each chosen candidate is marked as "in use" and the process
+	// continues, recursively. This way, all requests get matched against
+	// all candidates in all possible orders.
+	//
+	// The first full solution is chosen.
+	//
+	// In other words, this is an exhaustive search. This is okay because
+	// it aborts early. Once scoring gets added, more intelligence may be
+	// needed to avoid trying "equivalent" solutions (two identical
+	// requests, two identical devices, two solutions that are the same in
+	// practice).
+
+	// This is where we sanity check that we can actually handle the claims
+	// and their requests. For each claim we determine how many devices
+	// need to be allocated. If not all can be stored in the result, the
+	// claim cannot be allocated.
+	for claimIndex, claim := range alloc.claimsToAllocate {
+		numDevices := 0
+
+		// If we have any any request that wants "all" devices, we need to
+		// figure out how much "all" is. If some pool is incomplete, we stop
+		// here because allocation cannot succeed. Once we do scoring, we should
+		// stop in all cases, not just when "all" devices are needed, because
+		// pulling from an incomplete might not pick the best solution and it's
+		// better to wait. This does not matter yet as long the incomplete pool
+		// has some matching device.
+		for requestIndex := range claim.Spec.Devices.Requests {
+			request := &claim.Spec.Devices.Requests[requestIndex]
+			if request.DeviceClassName == "" {
+				// Unknown future request type!
+				return nil, fmt.Errorf("claim %s, request %s: unsupported request type", klog.KObj(claim), request.Name)
+			}
+			for i, selector := range request.Selectors {
+				if selector.CEL == nil {
+					// Unknown future selector type!
+					return nil, fmt.Errorf("claim %s, request %s, selector #%d: unsupported selector type", klog.KObj(claim), request.Name, i)
+				}
+			}
+
+			// Should be set. If it isn't, something changed and we should refuse to proceed.
+			if request.DeviceClassName == "" {
```

Comment:
Removed above because this one here is closer to the actual usage of the class name.

I also updated the error message a bit: it now describes what the client sees ("missing device class name") and then offers a suggestion about the root cause ("(unsupported request type?)").